src/main/scala/strawman/collection/immutable/Spandex.scala

```diff
@@ -0,0 +1,247 @@
+
+  override def flatMap[B](f: A => IterableOnce[B]): Spandex[B] =
+    fromIterable[B](View.FlatMap[A, B](coll, f))
+
+  override def partition(p: A => Boolean): (Spandex[A], Spandex[A]) = {
+    val pn = View.Partition(coll, p)
+    (fromIterable(pn.left), fromIterable(pn.right))
+  }
+
+  override def take(n: Int): Spandex[A] =
+    new Spandex.Secondary[A](primary, index, math.min(n, length))
+
+  override def drop(n: Int): Spandex[A] =
+    new Spandex.Secondary[A](primary, shift(math.min(n, length)), math.max(0, length - n))
+
+  override def tail: Spandex[A] =
+    if (length > 0) new Spandex.Secondary[A](primary, shift(1), length - 1)
+    else ???
+
+  override def reverse: Spandex[A] =
+    new Spandex.Secondary[A](primary, index, length, true)
+
+  override def iterator(): Iterator[A] = new Iterator[A] {
+    private[this] var i = 0
+    override def hasNext: Boolean = i < Spandex.this.length
+    override def next(): A = {
+      i += 1
+      fetch(i - 1)
+    }
+  }
+
+  def fromIterable[B](c: collection.Iterable[B]): Spandex[B] =
+    Spandex.fromIterable(c)
+
+  override def className = "Spandex"
+
+  override protected[this] def newBuilderWithSameElemType: Builder[A, Spandex[A]] = Spandex.newBuilder
+
+  override def newBuilder[E]: Builder[E, Spandex[E]] = Spandex.newBuilder
+}
+
+object Spandex extends IterableFactory[Spandex] {
+  private[immutable] class Primary[+A](
+      override val index: Int,
+      override val length: Int,
+      val array: scala.Array[Any])
+      extends Spandex[A] {
+    @volatile var low: Int = index
+    @volatile var high: Int = index + length - 1
+    override protected val primary: Primary[A] = this
+
+    override def +:[B >: A](b: B): Spandex[B] = {
+      if (lower(b, index, length)) new Secondary[B](this, index - 1, length + 1)
+      else b +: Spandex[A](index, length, array)
+    }
+    override def :+[B >: A](b: B): Spandex[B] = {
+      if (raise(b, index, length)) new Secondary[B](this, index, length + 1)
+      else Spandex[A](index, length, array) :+ b
+    }
+    private[Spandex] def lower[B >: A](b: B, i: Int, l: Int): Boolean = {
+      this.synchronized {
+        if (low > 0 && low == i) {
+          low -= 1
+          array(low) = b
+          true
+        } else false
+      }
+    }
+    private[Spandex] def raise[B >: A](b: B, i: Int, l: Int): Boolean = {
+      this.synchronized {
+        if (high < array.length - 1 && high == (i + l - 1)) {
+          high += 1
+          array(high) = b
+          true
+        } else false
+      }
+    }
+  }
+
+  private[immutable] final object Empty extends Primary[Nothing](0, 0, Array.empty) {
+    override def +:[B >: Nothing](b: B): Spandex[B] = Spandex(b)
+    override def :+[B >: Nothing](b: B): Spandex[B] = Spandex(b)
+    override def ++[B >: Nothing](xs: IterableOnce[B]): Spandex[B] = {
+      val builder = Spandex.newBuilder[B]
+      builder ++= xs
+      Spandex.fromIterable(builder.result)
```

Comment:
This is not the first time we face this situation: maybe our `fromIterable` method should be a `fromIterableOnce`, so that we avoid using the builder. Is there an issue in doing that?